serialization/src/tagged.rs

```diff
@@ -0,0 +1,213 @@
+//!     version: Tag<2>,
+//!     prev_hash: [u8; 32],
+//!     tx_root: [u8; 32],
+//!     time: u32,
+//!     signature: Vec<u8>,
+//! }
+//!
+//! // A top-level header type
+//! #[derive(DirectEncode, DirectDecode)]
+//! enum Header {
+//!     V1(HeaderV1),
+//!     v2(HeaderV2),
+//! }
+//!
+//! // Create a PoW header
+//! let header_v1 = HeaderV1 {
+//!     version: Tag,
+//!     prev_hash: [42; 32],
+//!     tx_root: [43; 32],
+//!     nonce: 177_984_342_498_312,
+//! };
+//! let header = Header::V1(header_v1.clone());
+//!
+//! // Serializing `Header` and `HeaderV1` gives the same encoding.
+//! assert_eq!(header.encode(), header_v1.encode());
+//! ```
+//!
+//! ## Illegal usage
+//!
+//! It is mandatory to specify the initial byte value, the following is not legal:
+//!
+//! ```compile_fail
+//! # use serialization::tagged::*;
+//! #[derive(Tagged)]
+//! enum Example { X }
+//! ```
+//!
+//! While this is:
+//!
+//! ```
+//! # use serialization::tagged::*;
+//! #[derive(Tagged)]
+//! enum Example { X = 5 }
+//! ```
+//!
+//! Tagged enums cannot have multiple variants:
+//! ```compile_fail
+//! # use serialization::tagged::*;
+//! #[derive(Tagged)]
+//! enum Example { X = 1, Y = 2 }
+//! ```
+//!
+//! Directly encoded enums have to have distinct tag values:
+//! ```compile_fail
+//! # use serialization::tagged::*;
+//! #[derive(DirectDecode)]
+//! enum Example { X(Tag<5>), Y(Tag<5>) }
+//! ```
+
+/// Derive the [Tagged] trait
+///
+/// * An `enum` with just one variant always starts with the variant index and can be made [Tagged].
+///   The `#[codec(index = X)]` or enum discriminant has to be provided explicitly.
+/// * A struct where the first element implements [Tagged] can be made Tagged.
+pub use serialization_derive::Tagged;
+
+/// Derive [Encode] that does not add the discriminant byte
+pub use serialization_derive::DirectEncode;
+
+/// Derive [Decode] for decoding without the initial discriminant byte
+pub use serialization_derive::DirectDecode;
+
+use crate::{Decode, Encode, Input};
+
+/// Types whose encoding starts with a fixed byte (tag)
+pub trait Tagged {
+    /// Initial byte in encoding of `Self`
+    const TAG: u8;
+}
+
+// Impls for some common wrapper types
+impl<T: Tagged> Tagged for &T {
+    const TAG: u8 = T::TAG;
+}
+impl<T: Tagged> Tagged for &mut T {
+    const TAG: u8 = T::TAG;
+}
+impl<T: Tagged> Tagged for Box<T> {
+    const TAG: u8 = T::TAG;
+}
+
+// Tuples are tagged if the first element is tagged
+macro_rules! impl_tagged_for_tuple {
+    ($T:ident, $($TS:ident),*) => {
+        impl<$T: Tagged, $($TS),*> Tagged for ($T, $($TS),*) {
+            const TAG: u8 = <$T as Tagged>::TAG;
+        }
+    };
+}
+// Up to 8 elements for now, can be expanded if needed
+impl_tagged_for_tuple!(T0,);
+impl_tagged_for_tuple!(T0, T1);
+impl_tagged_for_tuple!(T0, T1, T2);
+impl_tagged_for_tuple!(T0, T1, T2, T3);
+impl_tagged_for_tuple!(T0, T1, T2, T3, T4);
+impl_tagged_for_tuple!(T0, T1, T2, T3, T4, T6);
```

Comment:
Where is T5?

Reply:
Hmm maybe I should go back to elementary school. Fixing.